### ISSUE TITLE
POS Cash reconciliation 

### ISSUE THREAD

**Issue by ahmad18189** [frappe/erpnext#7582]
Hello ,
POS cash reconciliation for sales daily is very important feature for POS module .
in the end of day there should by a reconciliation for each POS employee where it show 
the sales invoice made by him and the actual cash or transaction.

please if some one have more details needed for this feature i am willing to develop it and make a pull request for it 
best regards
Ahmad

**Comment by mhbu50** [comment]
+1

**Comment by Olawale1** [comment]
+1

Very Very Important

**Comment by Olawale1** [comment]
Hi @ahmad18189 

Trust you're doing great. Any progress here?

Regards

**Comment by pifiu** [comment]
Please see this: https://github.com/frappe/erpnext/issues/2731

**Comment by xdelios** [comment]
I need this as well.
Any news on the progress?

**Comment by rmehta** [comment]
Start a bounty! Get the developers attention.

**Comment by heejit** [comment]
You can try following

1. We need to track start tran number and end tran number
   for each cashier.

2. We need to tag transaction as cash, card or credit

2. All sales will be debited to one customer a/c
   WALK-IN-CUSTOMER A/C

3. When you want to close the account 
   read all transaction for a cashier with start and end 
   transaction number 

4. Tally the cash, card and credit transaction

5. Post the JV
   CASH   DR
   CARD   DR
   CREDIT DR
     WALK-IN-CUSTOMER A/C CR
   
   End of the day walk-in-customer a/c must be zero 
   always.
   
6. Mark Start and end tran is closed for each cashier

**Comment by chdecultot** [comment]
Hi @ahmad18189 ,

Have you been able to finish your alpha version ?

I am also volunteering to help develop it.

**Comment by chdecultot** [comment]
Hi Everyone,

Since this topic seems to be on hold, I would like to propose the following and have started working on it.

I propose to divide the work in several parts. I didn't finish the spec for all steps, but I will update this ticket as soon as possible.

**Step 1**

In the scope of generating an X and a Z report, each transaction must be linked to a particular cashier.

My proposal is therefore to add a new field in the sales invoice Doctype, named for example "pos_profile" and to register the POS Profile making the transaction when it is submitted.

Every transaction should therefore be linked to a unique cashier.

It means that each POS Profile should only be applicable to one particular User. So we need to also change the POS Profile doctype to enable the parameter "Set Only Once" for a User.


I just pushed PR#9339 for this step.  


**Step 2**

Each new transaction should be flagged with a particular transaction number.
This transaction number could be the Offline_POS_Number and we could use it as a start and end transaction number.

It would require the addition of two more fields in the sales invoice Doctype: "pos_session_start" and "pos_session_end"


When the cashier starts it day, he/she would need to indicate the system that it is initializing the cash register.

I propose to add an extra layer on top of the existing one to make sure the cashier starts the new session when it opens the cash register:

![session_start_layer](https://user-images.githubusercontent.com/4903591/27233708-bdc26858-52ba-11e7-80ac-b066ee5f7fd2.png)

The "pos_session_start" can be filled with the name of the first invoice or by a timestamp.

Then we have several options:

1) Either add the extra layer everytime the cashier leaves the POS and come back with a message like "Resume Current Session"
2) Do not add any extra layer until the cashier has ended its session.


I would propose the second option since it simplifies the UI workflow.


At the end of the day, the cashier closes its register by clicking on "End Session":

![session_end_button 1](https://user-images.githubusercontent.com/4903591/27233728-ceb18996-52ba-11e7-94b3-3c4af301bdcf.png)



A confirmation message should pop-up to make sure the user didn't click by mistake.

After confirmation, the register should be closed and the "pos_session_end" should be updated with either the name of the last invoice or a timestamp.




**Step 3**


The above steps will make sure we have enough information to print an accurate X or Z report.


I propose to be add a field in the menu for the X report (report generated at any time during a session) and maybe a button or icon on the additional layer showed above to generate the Z report.

I will continue working on it and keep you updated.

**Comment by felixvarghese** [comment]
With regards to POS Profile, one idea may be the following. Fundamentally, a user will create a POS Profile mostly based on a location. So, if you have 5 locations and 10 users, that can be 50 profiles. However, what if you simply assigned users to a POS Profile? So, if you have 5 locations, create 5 profiles. Then assign users to those profiles.

This would be similar to assigning employees to salary structures. 

So, have a POS Profile, then have a child table of users. In the child table of users, you can assign users.

As for a unique identifier, each record in the user child table has a hash which can be used in Sales Invoice.

**Comment by chdecultot** [comment]
Hi @felixvarghese,

Yes that is a good idea! We should keep it in mind when working on #6308 :+1:

**Comment by Aintzanemasa** [comment]
Hi.

How about the cash reconciliation in pos? I think it's very important to the people who use de POS system in Erpnext. I think, there maybe a button wich show you how much money in cash do you have, and how much in the bank.
A report like this:

SALES DETAIL 20/07/2017

TOTAL AMOUN: 500€
CASH: 100€
CREDIT CARD: 400€

TAX

21% of 450 : 94.5
10% of 50: 5
4% of 0:

I think it's an easy report to include in the POS. and it will improve the use.

**Comment by rebin2002** [comment]
Hello, 

I have the same problem that I can not make multiple POS Profiles for 1 user, my case is as the following.

I'm distributing smart phones in shops, around 500 shops, and I have 10 representatives which means each Rep has 50 shops to control and visit them daily, we want to add these 500 shops as warehouses because we put our phones there and when its sold then they make payment otherwise we should return the goods which are not sold, that means when its sold to end user then will be an invoice.

So I need to make 50 POS profiles for each user so they can make quick invoice and receiving payment in each shop they visit for the phones which are sold.

In my case there is 2 solutions which need to be added in the system i think.

1. Allow Multiple POS Profiles per user.
2. In POS window there should be an option to select the Warehouse. 

Please if anyone has any other idea let me know. Thank you very much.

**Comment by jay-parikh** [comment]
Hello,
It looks like this issue is in dormant state. It would be great, if someone can provide any update / progress regarding this issue.

**Comment by rmehta** [comment]
@jay-parikh If its dormant, then clearly there is nothing to update 😉

**Comment by jay-parikh** [comment]
@rmehta  As it has been added in milestones: September 2017, October 2017 then again it has been removed. This issue is in Open state for quite a long time and looking at log trail seems that this has been pushed back for some or other reason. Just to know about what is it which keep this issue pushing back.

**Comment by nabinhait** [comment]
The report has been added into erpnext in develop branch named "Sales Payment Summary"
